### PR TITLE
feat(client): ✨ Add Receive Role

### LINKED ISSUE #30: Allow Catniss to receive a role
Upon game start, Catniss should receive either "cat" or "catfish" and it should display that role on the page so that she knows which "team" she is a part of and knows what to do during the game.

A hacker should not be able to view whether other players are a catfish because that would potentially ruin the entire lobby

- [x] Create role receiving screen
- [x] Create function to randomly assign roles
- [x] Make sure only the desired number of catfish is assigned in a lobby (make this configurable)

**Acceptance Criteria**

-  Once game starts, user is given a role either cat or catfish
- Displays the role to the user
- User shouldn't be able to see others role
- Hacker should not be able to see others role
- Hacker should not be able to influence who is cat and who is catfish

### PR DESCRIPTION
**Pull request by cmcclusky**
Adds the receive role function and component. Still needs lobby code to be imported and slight HTML changes.

### CODE REVIEW

File: client/src/components/ReceiveRole.svelte [hunk 1]
```diff
@@ -0,0 +1,74 @@
+<script lang="ts">
+  import type { PrivatePlayer } from "$lib/firebase/firestore-types/lobby";
+  import { getPrivatePlayerCollection } from "$lib/firebase/firestore-collections";
+  import { doc, setDoc, getDoc } from "firebase/firestore";
```
Comment (gchamb) [inline client/src/components/ReceiveRole.svelte]: setDoc isnt being used

File: client/src/lib/firebase/role.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      await setDoc(doc(getPrivatePlayerCollection(lobby), uids[j]), {
+        role: "CAT",
+        prompt: "",
+      });
+    } else {
+      await setDoc(doc(getPrivatePlayerCollection(lobby), uids[j]), {
+        role: "CATFISH",
+        prompt: "",
+      });
+    }
```
Comment (gchamb) [inline client/src/lib/firebase/role.ts:41]: if you're gonna set data from the client you need a firestore rule to validate it

File: client/src/lib/firebase/role.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+  // For loop that will create a random number between 0 and the amount of players.
+  // Then adds it to 'check', then loops back through again if needed, checking to see if the new random number was already selected and looping agian if so.
+  for (let i = 0; i < numCatFish; i++) {
+    current = Math.floor(Math.random() * (uids.length - 1) + 0);
+    if (check.indexOf(current) === -1) {
+      check.push(current);
+    } else i--;
+  }
+  // Loops through the players, checking to see if their spot in the user list matches with any of the random numbers.
```
Comment (gchamb) [inline client/src/lib/firebase/role.ts:28]: you dont want to decrement the looping counter

File: client/src/components/ReceiveRole.svelte [hunk 4]
```diff
@@ -0,0 +1,74 @@
+<script lang="ts">
```
Comment (gchamb) [inline client/src/components/ReceiveRole.svelte:1]: you need to merge with development to get the latest changes and test it again

File: client/src/components/ReceiveRole.svelte [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  import { assignRole } from "$lib/firebase/role";
+  import { authStore as user } from "$stores/auth";
+
+  export let code: string;
+  let getRole: PrivatePlayer;
+
+  // Runs the assignRole function using the lobby to assign every user a role. Then pulls the current user's uid and their data.
+  onMount(async () => {
+    const lobby = doc(lobbyCollection, code);
+    await assignRole(lobby);
```
Comment (gchamb) [inline client/src/components/ReceiveRole.svelte]: if I refreshed the page with this component could this change my role?

File: client/src/lib/firebase/role.ts (hunk 3 above)
Comment (nstringham) [reply client/src/lib/firebase/role.ts:28]: why not?

Comment (gchamb) [reply client/src/lib/firebase/role.ts:28]: i know some consider it bad practice

Comment (nstringham) [reply client/src/lib/firebase/role.ts:28]: how would you do it?

File: client/src/lib/firebase/role.ts (hunk 2 above)
Comment (nstringham) [reply client/src/lib/firebase/role.ts:41]: I think this should probobly be done in firebase fuctions for security reasons. you don't want hackers to be able to choose what role they are and you want the player to be the only one who can see whether or not they are a cat or a catfish.

**Review by nstringham — CHANGES_REQUESTED** [review]
see Giantte's comments

File: client/src/lib/firebase/role.ts (hunk 3 above)
Comment (nstringham) [reply client/src/lib/firebase/role.ts:28]: I guess you could use a nested do while loop

**Comment by SBAsai** [comment]
Updated pull request is [here](https://github.com/cat-acronym-team/cat-me-if-you-can/pull/109)

**Pull request closed without merging**